File: day8/day8.py
from itertools import combinations
from functools import reduce

def parse_input_into_matrix(filename):
    with open(filename) as f:
        lines = f.read().splitlines()
    return lines

def build_antenna_groups(grid):
    groups = {}
    for i in range(len(grid)):
        for j in range(len(grid[i])):
            val = grid[i][j]
            if val != '.':
                groups[val] = groups.get(val, []) + [(i, j)]

    return groups

def find_antinodes(group, max_x, max_y):
    antinodes = set()
    pairs = combinations(group, 2)
    for pair in pairs:
        first, second = sorted(pair)
        x_diff = second[0] - first[0]
        y_diff = second[1] - first[1]
        antinode_x1 = first[0] + x_diff * 2
        antinode_y1 = first[1] + y_diff * 2

        antinode_x2 = second[0] - x_diff * 2
        antinode_y2 = second[1] - y_diff * 2

        for x, y in [(antinode_x1, antinode_y1), (antinode_x2, antinode_y2)]:
            if 0 <= x < max_x and 0 <= y < max_y:
                antinodes.add((x, y))
    return antinodes

def find_resonant_antinodes(group, max_x, max_y):
    resonant_antinodes = set()
    pairs = combinations(group, 2)
    for pair in pairs:
        first, second = sorted(pair)
        x_diff = second[0] - first[0]
        y_diff = second[1] - first[1]
        resonant_antinode_x = first[0] + x_diff
        resonant_antinode_y = first[1] + y_diff
        while 0 <= resonant_antinode_x < max_x and 0 <= resonant_antinode_y < max_y:
            resonant_antinodes.add((resonant_antinode_x, resonant_antinode_y))
            resonant_antinode_x += x_diff
            resonant_antinode_y += y_diff

        resonant_antinode_x = second[0] - x_diff
        resonant_antinode_y = second[1] - y_diff
        while 0 <= resonant_antinode_x < max_x and 0 <= resonant_antinode_y < max_y:
            resonant_antinodes.add((resonant_antinode_x, resonant_antinode_y))
            resonant_antinode_x -= x_diff
            resonant_antinode_y -= y_diff
    return resonant_antinodes

def part1(filename):
    input_data = parse_input_into_matrix(filename)
    groups = build_antenna_groups(input_data)
    unique_antinodes = reduce(lambda x, y: x.union(y), [find_antinodes(groups[group], len(input_data), len(input_data[0])) for group in groups])
    return len(unique_antinodes)

def part2(filename):
    input_data = parse_input_into_matrix(filename)
    groups = build_antenna_groups(input_data)
    unique_resonant_antinodes = reduce(lambda x, y: x.union(y), [find_resonant_antinodes(groups[group], len(input_data), len(input_data[0])) for group in groups])
    return len(unique_resonant_antinodes)

input_data = parse_input_into_matrix('test.txt')
# print(input_data)

groups = build_antenna_groups(input_data)
# print(build_antenna_groups(input_data))

# print(find_antinodes(groups['A'], len(input_data), len(input_data[0])))

# print(part1('real.txt'))
print(part2('real.txt'))

File: day8/day8.clj
(ns day8)

(defn parse_input [filename]
  (clojure.string/split (slurp filename) #"\n")
  )

(defn get-index [item coll]
  (remove nil?
          (map-indexed (fn [ix x] 
                         (let [i (clojure.string/index-of x item)]
                                    (when i [ix i])
                                    )
                         )
                       coll
                       )
          )
  )

(defn find_antenna_groups [input]
  (let [unique_antennas (filter (fn [x] (not= \. x)) (distinct (reduce str input)))]

  (->>
   (map (fn [antenna] {antenna (get-index antenna input)}) unique_antennas)
   (reduce merge {})
   )
  )
  )

(defn single_mapper [ant1 ant2]
  (let [x1 (first ant1)
        y1 (second ant1)
        x2 (first ant2)
        y2 (second ant2)
        xdiff (- x2 x1)
        ydiff (- y2 y1)
        ]
    [
     [(+ x1 (* 2 xdiff)) (+ y1 (* 2 ydiff))]
     [(- x2 (* 2 xdiff)) (- y2 (* 2 ydiff))]
     ]
    )
  )

(defn resonant_mapper [ant1 ant2 max_x max_y]
  (let [x1 (first ant1)
        y1 (second ant1)
        x2 (first ant2)
        y2 (second ant2)
        xdiff (- x2 x1)
        ydiff (- y2 y1)
        ]
    (def node1 (atom [x1 y1]))
    (def node2 (atom [x2 y2]))
    (def nodes (atom []))

    (while (and
            (<= 0 (first @node1) max_x)
            (<= 0 (second @node1) max_y)
            ) 
      (swap! nodes conj [(first @node1) (second @node1)])
      (swap! node1 
             (fn [node] [(+ (first node) xdiff) (+ (second node) ydiff)]) 
)
      )
    (while (and
            (<= 0 (first @node2) max_x)
            (<= 0 (second @node2) max_y)
            ) 
      (swap! nodes conj [(first @node2) (second @node2)])
      (swap! node2 
             (fn [node] [(- (first node) xdiff) (- (second node) ydiff)])
      )
    )
    @nodes
    ) 
  )

(defn find_antinodes [group max_x max_y mapper resonant?]
  (let [pairs (for [x group
                    y group
                    :when (not= x y)]
               [x y])
       ]
    (->> pairs
         (map (fn [[ant1 ant2]]
                (if resonant?
                  (mapper ant1 ant2 max_x max_y)
                  (mapper ant1 ant2)
                )
                )
              )
         (reduce concat)
         (distinct)
         (filter (fn [[x y]]
                   (and
                    (<= 0 x max_x)
                    (<= 0 y max_y)
                    )
                   )
                 )
         )
  )
)


(defn part1 [filename]
  (let [input (parse_input filename)
        max_x (dec (count (first input)))
        max_y (dec (count input))
        groups (find_antenna_groups input)
        ]
    (->> groups
         (vals)
         (map (fn [group] (find_antinodes group max_x max_y single_mapper false))) 
         (reduce concat)
         (distinct)
         (count)
         )
     )
  )

(defn part2 [filename]
  (let [input (parse_input filename)
        max_x (dec (count (first input)))
        max_y (dec (count input))
        groups (find_antenna_groups input)]
    (->> groups
         (vals)
         (map (fn [group] (find_antinodes group max_x max_y resonant_mapper true)))
         (reduce concat)
         (distinct)
         (count)
         )
    )
  )